File: netlify/functions/storage-recommendation.py
import os
import json
from openai import OpenAI

HEADERS = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Headers": "Content-Type",
    "Content-Type": "application/json",
}

def handler(event, context):
    # السماح لطلبات الـ OPTIONS (preflight)
    if event.get("httpMethod") == "OPTIONS":
        return {"statusCode": 200, "headers": HEADERS, "body": ""}

    if event.get("httpMethod") != "POST":
        return {"statusCode": 405, "headers": HEADERS,
                "body": json.dumps({"error": "Method Not Allowed"})}

    try:
        raw_body = event.get("body") or "{}"
        body = json.loads(raw_body)
    except json.JSONDecodeError:
        return {"statusCode": 400, "headers": HEADERS,
                "body": json.dumps({"error": "Invalid JSON body"})}

    user_requirements = body.get("requirements", "").strip()
    language = body.get("language", "en").strip().lower()

    if not user_requirements:
        return {"statusCode": 400, "headers": HEADERS,
                "body": json.dumps({"error": "Requirements field is required"})}

    api_key = os.getenv("OPENAI_API_KEY")
    if not api_key:
        return {"statusCode": 500, "headers": HEADERS,
                "body": json.dumps({"error": "Missing OPENAI_API_KEY"})}

    # تهيئة العميل
    client = OpenAI(api_key=api_key)

    # البرومبت
    if language == "tr":
        system_prompt = (
            "Sen bir depolama cihazı uzmanısın. Kullanıcının ihtiyaçlarına göre en uygun depolama çözümünü öner.\n"
            "Yanıtın şunları içersin:\n"
            "1) Tür (SSD/HDD/NVMe)\n2) Kapasite\n3) Türkiye pazarı için tahmini fiyat aralığı (TRY)\n"
            "4) Önerilen markalar\n5) Neden uygun olduğunun kısa açıklaması.\n"
            "Kısa, net ve anlaşılır yaz."
        )
        user_prompt = f"Kullanıcı ihtiyacı: {user_requirements}"
    else:
        system_prompt = (
            "You are a storage expert. Recommend the most suitable storage solution based on the user's needs.\n"
            "Include: 1) Type (SSD/HDD/NVMe) 2) Capacity 3) Estimated price range "
            "(default TRY for Turkey; use USD otherwise) 4) Recommended brands 5) Brief rationale.\n"
            "Be concise and practical."
        )
        user_prompt = f"User requirements: {user_requirements}"

    try:
        resp = client.chat.completions.create(
            model="gpt-4o-mini",          # متوافق مع chat.completions
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt},
            ],
            max_tokens=420,               # تقليل بسيط لتفادي المهلة
            temperature=0.7,
        )
        recommendation = resp.choices[0].message.content.strip()

        return {"statusCode": 200, "headers": HEADERS,
                "body": json.dumps({"success": True,
                                    "recommendation": recommendation,
                                    "language": language})}

    except Exception as e:
        msg = str(e)
        code = 500
        if "authentication" in msg.lower():
            code = 401
        elif "rate limit" in msg.lower():
            code = 429
        return {"statusCode": code, "headers": HEADERS,
                "body": json.dumps({"error": f"Internal server error: {msg}"})}

File: netlify/functions/storage-recommendation.js
// Netlify Function (Node, CommonJS) — /.netlify/functions/storage-recommendation
const OpenAI = require("openai");

const HEADERS = {
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Headers": "Content-Type",
  "Access-Control-Allow-Methods": "POST, OPTIONS",
  "Content-Type": "application/json",
};

exports.handler = async (event) => {
  // CORS preflight
  if (event.httpMethod === "OPTIONS") {
    return { statusCode: 200, headers: HEADERS, body: "" };
  }

  // Enforce POST
  if (event.httpMethod !== "POST") {
    return {
      statusCode: 405,
      headers: HEADERS,
      body: JSON.stringify({ error: "Method Not Allowed" }),
    };
  }

  // Parse body safely
  let body;
  try {
    body = JSON.parse(event.body || "{}");
  } catch {
    return {
      statusCode: 400,
      headers: HEADERS,
      body: JSON.stringify({ error: "Invalid JSON body" }),
    };
  }

  const requirements = String(body.requirements || "").trim();
  const language = String(body.language || "en").trim().toLowerCase();

  if (!requirements) {
    return {
      statusCode: 400,
      headers: HEADERS,
      body: JSON.stringify({ error: "Requirements field is required" }),
    };
  }

  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) {
    // لا تُظهر سبب داخلي للمستخدم النهائي
    return {
      statusCode: 500,
      headers: HEADERS,
      body: JSON.stringify({ error: "Service temporarily unavailable" }),
    };
  }

  // Optional: short timeout via AbortController
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 25_000);

  try {
    const client = new OpenAI({ apiKey });

    // Avoid multi-line template literals to prevent syntax issues in deployment.
    const system =
      language === "tr"
        ? [
            "Sen bir depolama cihazı uzmanısın.",
            "ÇIKTIYI SADECE Markdown olarak ver (HTML yok).",
            "Aşağıdaki formatı AYNEN uygula (başlıklar ve sıralama dahil):",
            "",
            "## Önerilen Depolama Çözümü",
            "",
            "1. **Type:** ...",
            "2. **Capacity:** ...",
            "3. **Estimated Price Range (TRY):** ...",
            "4. **Recommended Brands:** ...",
            "5. **Brief Rationale:** ...",
            "",
            "Kurallar:",
            "- Her madde mutlaka olsun, boş bırakma.",
            "- Fiyatı Türkiye için TRY olarak ver (yaklaşık aralık).",
            "- Markaları virgülle ayır.",
            "- Kısa, net, iş odaklı yaz.",
          ].join("\n")
        : [
            "You are a storage expert.",
            "OUTPUT MUST BE Markdown only (no HTML).",
            "Follow this EXACT format (including headings and order):",
            "",
            "## Recommended Storage Solution",
            "",
            "1. **Type:** ...",
            "2. **Capacity:** ...",
            "3. **Estimated Price Range (TRY):** ...",
            "4. **Recommended Brands:** ...",
            "5. **Brief Rationale:** ...",
            "",
            "Rules:",
            "- Include every item; do not omit any.",
            "- Price must be in TRY (approximate range) for Turkey.",
            "- Separate brands with commas.",
            "- Keep it concise, practical, business-focused.",
          ].join("\n");

    const userPrompt =
      language === "tr"
        ? `Kullanıcı ihtiyacı: ${requirements}`
        : `User requirements: ${requirements}`;

    const resp = await client.chat.completions.create(
      {
        model: "gpt-4o-mini",
        messages: [
          { role: "system", content: system },
          { role: "user", content: userPrompt },
        ],
        max_tokens: 420,
        temperature: 0.7,
      },
      { signal: controller.signal }
    );

    const recommendation = String(resp.choices?.[0]?.message?.content || "").trim();

    return {
      statusCode: 200,
      headers: HEADERS,
      body: JSON.stringify({ success: true, recommendation, language }),
    };
  } catch (e) {
    // Log for you (Netlify logs) without exposing internals to the public
    console.error("storage-recommendation error:", e);

    // لا تُظهر السبب الحقيقي للجمهور
    return {
      statusCode: 500,
      headers: HEADERS,
      body: JSON.stringify({ error: "Service temporarily unavailable" }),
    };
  } finally {
    clearTimeout(timeout);
  }
};
